src/symmetric.rs

```diff
@@ -471,6 +643,94 @@ impl SymmetricAlgorithmKey {
     }
 }
 
+#[cfg(feature = "block-cipher-trait")]
+pub use block_cipher;
+#[cfg(feature = "block-cipher-trait")]
+mod block_cipher_trait {
+    use super::*;
+
+    use core::convert::TryFrom;
+
+    use block_cipher::generic_array::{
+        typenum::{self, Unsigned},
+        ArrayLength,
+    };
+    use block_cipher::{self, Block, BlockCipher, Key, NewBlockCipher};
+
+    impl KeyBits for typenum::U128 {
+        const VALUE: Option<usize> = Some(Self::USIZE);
+    }
+    impl KeyBits for typenum::U192 {
+        const VALUE: Option<usize> = Some(Self::USIZE);
+    }
+    impl KeyBits for typenum::U256 {
+        const VALUE: Option<usize> = Some(Self::USIZE);
+    }
+
+    #[test]
+    fn assert_aes_keysize_valid() {
+        fn _inner<T: NewBlockCipher>() {}
+        _inner::<super::Key<Aes, typenum::U128>>();
+        _inner::<super::Key<Aes, typenum::U192>>();
+        _inner::<super::Key<Aes, typenum::U256>>();
+    }
+
+    impl<B: KeyBits> NewBlockCipher for super::Key<Aes, B>
+    where
+        B: typenum::Unsigned,
+        // Fancy way of allowing only {128, 192, 256}
+        B: typenum::IsGreaterOrEqual<typenum::U128, Output = typenum::B1>,
+        B: typenum::IsLessOrEqual<typenum::U256, Output = typenum::B1>,
+        B: typenum::PartialDiv<typenum::U64>,
+        // Help the trait solver see that it's also divisible by 8
+        B: typenum::PartialDiv<typenum::U8>,
+        <B as typenum::PartialDiv<typenum::U8>>::Output: ArrayLength<u8>,
+    {
+        /// Key size in bytes with which cipher guaranteed to be initialized.
+        type KeySize = <B as typenum::PartialDiv<typenum::U8>>::Output;
+
+        /// Create new block cipher instance from key with fixed size.
+        fn new(key: &Key<Self>) -> Self {
+            let prov =
+                SymmetricAlgorithm::open(SymmetricAlgorithmId::Aes, ChainingMode::Ecb).unwrap();
+            let key = prov.new_key(key).unwrap();
+            match Self::try_from(key) {
+                Ok(value) => value,
+                Err(..) => panic!(),
+            }
+        }
+    }
+
+    impl<B: KeyBits> BlockCipher for super::Key<Aes, B> {
+        /// Size of the block in bytes
+        type BlockSize = typenum::U16;
+
+        /// Number of blocks which can be processed in parallel by
+        /// cipher implementation
+        type ParBlocks = typenum::U1;
+
+        /// Encrypt block in-place
+        fn encrypt_block(&self, block: &mut Block<Self>) {
+            // FIXME: We assume here that we use ECB (as initialized via the
+            // NewBlockCipher trait)
```

Comment:
I don't get how the `BlockCipher` will decide what chaining mode to use?

Reply:
See https://github.com/emgre/win-crypto-ng/pull/31/files#r503954069

Reply:
I forgot about that we fact that we can cast `SymmetricAlgorithmKey` to Key - that's where using other chaining modes than ECB will not be possible...

One solution is to encode chaining mode alongside `KeyBits` as a generic parameter and only then `impl BlockCipher for Key<Aes/* for now */, Ecb, {128,192,256}>` but I'm worried about encoding more and more details in the compile-time context.

The other is to either put a friendly warning that only ECB keys support `BlockCipher` and let it blow up or `panic!` ourselves if the chaining mode for the key used for `encrypt_block`/`decrypt_block` is not in ECB mode.

Do you have any idea and/or preference how we should handle this?

Reply:
I'm not convinced by the extra type, especially since it would only be used to check that it's ECB. Seems a bit complicated for not much. Perhaps the `panic!` is the best option. We could panic in the `encrypt_block` / `decrypt_block` or straight in the `TryFrom`.

Reply:
I'm not 100% sure either is a good location, unfortunately:
- `*_block` operations will be called somewhat often and so it'd be great to check if we can panic only once
- If we intend to have a general compile-time known handle to a symmetric key (which it should be perfectly valid and even preferred to have a different chaining mode than ECB) then we shouldn't panic in `TryFrom` only to have a useful trait implemented.

Maybe having a helper struct similar to [`std::path::Display`](https://doc.rust-lang.org/std/path/struct.Display.html) would be a correct choice? If we couple it with a fallible conversion, then we can only do the check once

Reply:
I implemented that approach in a9c9aaa, let me know what you think.